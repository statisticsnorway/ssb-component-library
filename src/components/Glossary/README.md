Glossary
========

> Pop-up box for explaining a word

### Usage

#### HTML
Toggle the class "open" using javascript
```html
<button class="ssb-glossary"><!-- click to toggle 'open' class -->
  <div class="glossary-text-wrap">Explain this</div>
  <i>{feather.openBook 12px}</i>
  <div class="glossary-animate-background"></div>
  <div class="ssb-glossary-popup {open or closed}">
    <div class="content-box">
      <span class="info-text">{insert explanation}</span>
      <div class="ssb-glossary-closing">
        <i class="icon">{feather.xCircle 14px}</i>
        <span>Lukk</span>
      </div>
    </div>
  </div>
</button>
```

#### React

```jsx harmony
<Glossary explanation={placeholderText}>Explain this</Glossary>
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| children | node | Something to wrap the click event around |
| explanation | required string | Text to be inside popup |
