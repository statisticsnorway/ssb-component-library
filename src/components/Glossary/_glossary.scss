.ssb-glossary {
  @include focus-marker;
  @include reset-button;
  cursor: pointer;
  display: inline-block;
  position: relative;

  .glossary-text-wrap {
    border-bottom: 1px dotted $ssb-green-4;
    display: inline-block;
    line-height: inherit;
  }

  .glossary-logo {
    color: $ssb-green-3;
    margin: 2px;
    transition: color .2s;
    vertical-align: top;
  }

  .glossary-animate-background {
    background: $ssb-green-2;
    bottom: 0;
    height: 0;
    left: 0;
    position: absolute;
    transition: height .2s;
    width: 100%;
    z-index: -1;
  }

  &:hover, &:focus {

    .glossary-animate-background {
      height: 100%;
    }

    .glossary-logo {
      color: $ssb-dark-6;
    }
  }

  .glossary-popup {
    @include open-sans;
    animation: fadein .2s;
    display: none;
    font-size: 14px;
    left: 50%;
    margin-left: -135px;
    opacity: 0;
    position: absolute;
    top: 34px;
    transition: opacity .18s;
    width: 270px;

    &:after {
      border-color: transparent transparent $ssb-dark-5 transparent;
      border-style: solid;
      border-width: 12px;
      content: '';
      left: 50%;
      margin-left: -12px;
      position: absolute;
      top: -22px;
    }

    &.open {
      display: inline-block;
      opacity: 1;
    }

    .content-box {
      background: $ssb-dark-5;
      color: $ssb-white;
      padding: 20px;

      .ssb-glossary-closing {
        align-items: center;
        display: flex;
        justify-content: center;
        margin-top: 10px;
        user-select: none;
        width: 100%;

        .icon {
          color: $ssb-dark-5;
          fill: $ssb-green-2;
          font-size: 16px;
        }

        span {
          line-height: 1.43;
          margin-left: 2px;
        }
      }

      .info-text {
        color: $ssb-white;
        font-size: 14px;
      }
    }
  }
}
