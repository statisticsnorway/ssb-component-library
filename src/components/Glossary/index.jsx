import React, { useEffect, useState, useRef } from 'react';
import PropTypes from 'prop-types';
import { BookOpen, XCircle } from 'react-feather';

const Glossary = ({ explanation, children, closeText }) => {
	const node = useRef();
	const infoContainer = useRef();
	const [open, setOpen] = useState(false);

	const handleClickOutside = e => {
		if (node.current.contains(e.target)) {
			return;
		}
		setOpen(false);
	};

	useEffect(() => {
		if (open) {
			document.addEventListener('mousedown', handleClickOutside);
		} else {
			document.removeEventListener('mousedown', handleClickOutside);
		}

		return () => {
			document.removeEventListener('mousedown', handleClickOutside);
		};
	}, [open]);

	return (
		<button ref={node} onClick={() => setOpen(!open)} className="ssb-glossary">
			<div className="glossary-text-wrap">{children}</div>
			<BookOpen size={12} className="glossary-logo" />
			<div className="glossary-animate-background" />
			{open && (
				<div className={`glossary-popup${open ? ' open' : ''}`} ref={infoContainer}>
					<div className="content-box">
						<span className="info-text">{explanation}</span>
						<div className="ssb-glossary-closing">
							<XCircle size={16} className="icon" /><span>{closeText}</span>
						</div>
					</div>
				</div>
			)}

		</button>
	);
};

Glossary.defaultProps = {
	closeText: 'Lukk',
};

Glossary.propTypes = {
	children: PropTypes.node,
	closeText: PropTypes.string,
	explanation: PropTypes.string.isRequired,
};

export default Glossary;
