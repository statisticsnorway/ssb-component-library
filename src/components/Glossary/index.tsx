import React, { useEffect, useState, useRef, useCallback, type ReactNode } from 'react'
import { BookOpen, XCircle } from 'react-feather'
import classNames from '../../utils/utils'

interface GlossaryProps {
  children?: ReactNode
  className?: string
  closeText?: string
  explanation: string
  ariaLabel?: string
}

const Glossary: React.FC<GlossaryProps> = ({ explanation, children, className, closeText = 'Lukk', ariaLabel }) => {
  const node = useRef<HTMLSpanElement>(null)
  const [open, setOpen] = useState(false)

  const handleClickOutside = (e: MouseEvent) => {
    if (node.current?.contains(e.target as Node)) {
      return
    }
    setOpen(false)
  }

  const handleCloseButton = () => {
    setOpen(false)
    node.current?.focus()
  }

  const handleKeyDownClose = (e: React.KeyboardEvent<HTMLButtonElement>) => {
    if (e.key === 'Enter' || e.key === ' ') {
      e.preventDefault()
      setOpen(false)
      node.current?.focus()
    }
  }

  const escKeyListener = useCallback((e: KeyboardEvent) => {
    if (e.key === 'Escape') {
      setOpen(false)
      node.current?.focus()
    }
  }, [])

  useEffect(() => {
    if (open) {
      document.addEventListener('mousedown', handleClickOutside)
      document.addEventListener('keydown', escKeyListener, false)
    }

    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
      document.removeEventListener('keydown', escKeyListener, false)
    }
  }, [open])

  return (
    <span className={classNames('ssb-glossary', className)}>
      <button
        onClick={() => setOpen(!open)}
        className='glossary-button'
        aria-label={ariaLabel || (typeof children === 'string' ? children : undefined)}
        aria-expanded={open ? 'true' : 'false'}
      >
        <span className='glossary-text-wrap'>{children}</span>
        <BookOpen size={12} className='glossary-logo' aria-hidden='true' />
      </button>
      <span className={classNames('glossary-popup', open && 'open')} ref={node}>
        <span className='content-box'>
          <span className='info-text'>{explanation}</span>
          <span className='glossary-closing'>
            <button
              onClick={() => {
                handleCloseButton()
              }}
              onKeyDown={(e) => handleKeyDownClose(e)}
            >
              <XCircle size={16} className='icon' aria-hidden='true' />
              <span>{closeText}</span>
            </button>
          </span>
        </span>
      </span>
    </span>
  )
}

export default Glossary
