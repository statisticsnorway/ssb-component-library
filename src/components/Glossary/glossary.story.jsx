import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import Glossary from './index';
import Text from '../Text';

const placeholderText = `This is the explanation of the word
written in one or more paragraphs depending on the amount of space needed`;

storiesOf('Glossary', module).addDecorator(centered)
	.add('As a wrapper', () => (
		<Text>You can explain a single <Glossary explanation={placeholderText} closeText="Close">word</Glossary>
			&nbsp;in a sentence.
		</Text>
	));
