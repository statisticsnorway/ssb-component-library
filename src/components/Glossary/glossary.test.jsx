import React from 'react';
import { shallow } from 'enzyme';
import Glossary from './index';

describe('Glossary', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(<Glossary explanation="placeholderText">Glossary</Glossary>);
		expect(wrapper).toMatchSnapshot();
	});
});