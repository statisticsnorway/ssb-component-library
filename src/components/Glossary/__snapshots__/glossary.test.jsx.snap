// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Glossary Matches the snapshot 1`] = `
<button
  className="ssb-glossary"
  onClick={[Function]}
>
  <div
    className="glossary-text-wrap"
  >
    Glossary
  </div>
  <BookOpen
    className="glossary-logo"
    color="currentColor"
    size={12}
  />
  <div
    className="glossary-animate-background"
  />
</button>
`;
